### ISSUE TITLE
Range of dates with infinite end are stored incorrectly in Postgres database

### ISSUE THREAD

**Issue by pean** [rails/rails#51745]
In Rails 7.1.3 date ranges with infinite end is stored incorrectly when the end is included and it is i stored with excluded end.

### Steps to reproduce

1. Create a table with a Postgres date range column
2. Create and entry in the table with a range with included infinite end `(Date.today..)` and save it
3. Reload that entry and the range comes back with excluded end `(Date.today...)`

```ruby
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", '~> 7.1.3'
  gem "pg"
end

require "active_record"
require "minitest/autorun"
require "logger"

# You may need to change your pg settings accordingly
db_config = {
  adapter: 'postgresql',
  database: 'infinityrange_test',
  host: 'localhost', # Change this to your PostgreSQL host
  port: 5432,
  username: 'postgres', # Change this to your PostgreSQL username
  password: 'postgres'  # Change this to your PostgreSQL password
}

begin
  ActiveRecord::Base.establish_connection(db_config.except(:database))
  ActiveRecord::Base.connection.drop_database(db_config[:database]) rescue nil
  ActiveRecord::Base.connection.create_database(db_config[:database])
end
ActiveRecord::Base.establish_connection(db_config)
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :infinity_ranges do |t|
    t.tsrange :date_range
  end
end

class InfinityRange < ActiveRecord::Base; end

class InfinityRangeTest < Minitest::Test
  def test_date_range_infinity_
    # Create a range with infinity end including end 2024-01-01 00:00:00 UTC..
    # There are multiple ways to create a range with infinity end but all of 
    # these give the same result
    # range = (DateTime.new(2024)..Float::INFINITY)
    # range = (DateTime.new(2024)..DateTime::Infinity.new)
    # range = (DateTime.new(2024)..Date::Infinity.new)
    range = (DateTime.new(2024)..)

    infinity_range = InfinityRange.new(date_range: range)

    # Range is including end
    refute_predicate infinity_range.date_range, :exclude_end?

    # Save an reload the record to read back the value from db
    infinity_range.save!
    infinity_range.reload

    # Range is no longer including range and returns 2024-01-01 00:00:00 UTC...
    # but it should be 2024-01-01 00:00:00 UTC..
    assert_equal (DateTime.new(2024)..), infinity_range.date_range
    refute_predicate infinity_range.date_range, :exclude_end?

    # To read back excluding end it needs to be stored likt this
    assert_equal(
      "[#{DateTime.new(2024)},infinity]",
      infinity_range.read_attribute_before_type_cast(:date_range),
    )
  end

  # This test is to show that the range can be inserted directly into the database
  def test_direct_insert
    # Insert the equivalent of (DateTime.new(2024)..) directly into the database
    sql = "INSERT INTO infinity_ranges (date_range) VALUES ('[#{DateTime.new(2024)},infinity]')"
    ActiveRecord::Base.connection.execute(sql)

    date_range = InfinityRange.last.date_range
    assert_equal (DateTime.new(2024)..), date_range
    refute_predicate date_range, :exclude_end?
  end
end

```

### Expected behavior
To save the range including end as it was specified.

### Actual behavior
Range is saved in database excluding end and when it's read back into rails it is not the same value.

### System configuration
**Rails version**: `7.1.3`

**Ruby version**: `ruby 3.2.3 (2024-01-18 revision 52bb2ac0a6) [arm64-darwin23]`

**Comment by Earlopain** [comment]
Duplicate/Related to #51111?

**Comment by pean** [comment]
> Duplicate/Related to #51111?

Not really, that is about how data is read back from the database and the change that unbounded values now comes back as `nil` where the before Rails 7.1 came back as `Float::INFINITY`, which I think is a change to the better.

This is about what how the data is going into the database. When a boundless end is sent to PostgreSQL it is sent as `nil` which in PostgreSQL will be stored as an excluded end, but if it is sent as `infinity` it will be stored as inclusive, and be returned as such.

It is possible that this could be considered a PostgreSQL problem too,  because `'range = ["2024-01-01 00:00",infinity]` could be the same as `'range = ["2024-01-01 00:00",]`, but the latter is saved as `["2024-01-01 00:00",)`.

From what I can tell, the value that goes into the database has to be modified from `nil`, to string `infinity`. But I could be wrong.

**Comment by justinko** [comment]
It's inserting as `"[2024-01-01 00:00:00,]"`, and it looks like [Postgres will convert](https://www.postgresql.org/docs/current/rangetypes.html#RANGETYPES-INFINITE) it to `"[2024-01-01 00:00:00,)"` (exclusive).

Looks like we have to be explicit with the `infinity`.

**Comment by justinko** [comment]
All of these ranges in your test case:

```
    # range = (@time..Float::INFINITY)
    # range = (@time..DateTime::Infinity.new)
    # range = (@time..Date::Infinity.new)
```

Will pass with this change:

<img width="524" alt="image" src="https://github.com/rails/rails/assets/14007/41d35bd0-f58e-4ae9-93b7-bbd2bd596183">

I'm still investigating if this change is the right approach ... haven't even run the rails tests yet. 

If we detect if the range is a date/time range, then we could have `(@time..)` with `infinity` ... but then you get into "special case land" and I'm not sure that's a path to go down.

**Comment by pean** [comment]
> It's inserting as `"[2024-01-01 00:00:00,]"`, and it looks like [Postgres will convert](https://www.postgresql.org/docs/current/rangetypes.html#RANGETYPES-INFINITE) it to `"[2024-01-01 00:00:00,)"` (exclusive).
> 
> Looks like we have to be explicit with the `infinity`.

Exactly, that is true, they will convert like that, but if inserting the upper bound as `infinity` will give a more expected result (imo).

> I'm still investigating if this change is the right approach ... haven't even run the rails tests yet.
>
> If we detect if the range is a date/time range, then we could have (@time..) with infinity ... but then you get into "special case land" and I'm not sure that's a path to go down.

Make the same change while investigating this and to see if I could figure out a good solution, but I also quickly arrived to "special case land". But since this is a very specific case in just related to the postgres adapter maybe that can work anyway. 

I haven't yet been able to come up with a clean enough solution, got stuck with range value validation as this becomes a range with a mix of types dates and strings and it got even more into special case land 😃 

Not sure if it at this point worth to put up a work in progress PR that we can discuss around and if that is an acceptable way of working with this to move foward. This is the first time I'm digging this deep into rails code. Fun stuff! 😄

**Comment by justinko** [comment]
What do you think about this? https://github.com/rails/rails/compare/main...justinko:rails:issue-51745
